### PR TITLE
TopLoc Proof Generation

### PR DESCRIPTION
**Pull request by Jackmin801**

### CODE REVIEW

File: src/genesys/data.py [hunk 1]
```diff
@@ -72,6 +72,7 @@ def _add_column(dataset, path):
 
         self.total_samples = min(max_samples, total_samples)
 
+        self.tokenize = tokenize
         self.tokenizer = tokenizer
```
Comment (samsja) [inline src/genesys/data.py:76]: lets differentiate both variable. tokenize and tokenizer are difficult to put apart

File: src/genesys/generate.py [hunk 2]
```diff
@@ -82,20 +85,28 @@ def main(config: GenerateConfig):
...
+            )
             all_results.append(batch_element)
         total_samples += len(batch)
 
         if len(all_results) >= config.sample_per_file:
             file_name = f"{config.out_file_prefix}_{uuid.uuid4()}.jsonl"
             file = os.path.join(config.path_output, file_name)
             save_batch_results(all_results, file, gcp_bucket)
+            file_sha = sha256sum(file)
+            dataloader.prime_metric.log_prime({"file_sha": file_sha, "file_name": file_name})
```
Comment (samsja) [inline src/genesys/generate.py:108]: I would refactor and instantiate the prime metric in the main script and pass it as param to the dataalodaer.

ideally the prime metric should be as simple to use as wandb

File: src/genesys/toploc/__init__.py [hunk 3]
```diff
@@ -0,0 +1,2 @@
+from .poly import ProofPoly, build_proofs, build_proofs_base64  # noqa: F401
```
Comment (samsja) [inline src/genesys/toploc/__init__.py:1]:
```suggestion
from genesys.toploc.poly import ProofPoly, build_proofs, build_proofs_base64  # noqa: F401
```

lets use absolute import ?

Comment (samsja) [reply src/genesys/toploc/__init__.py:1]: Don't you think that it makes sense to create a separate package for toploc ? 

or maybe we incub it here for now

File: src/genesys/utils.py [hunk 4]
```diff
@@ -21,6 +22,9 @@
 from huggingface_hub import snapshot_download
 from datasets import load_dataset
 
+if TYPE_CHECKING:
+    from pathlib import Path
```
Comment (samsja) [inline src/genesys/utils.py:26]: its fine to import pathlib even if you just use as type hint

File: tests/toploc/test_poly.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
+import pytest
+import torch
+import base64
+from genesys.toploc.poly import (
+    find_injective_modulus,
+    build_proofs,
+    build_proofs_base64,
+    ProofPoly,
+)
+
```
Comment (samsja) [inline tests/toploc/test_poly.py:10]: test I love test

File: src/genesys/toploc/__init__.py (hunk 3 above)
Comment (Jackmin801) [reply src/genesys/toploc/__init__.py:1]: Yea I think it makes sense to be a separate package. the validator uses it too

File: src/genesys/utils.py (hunk 4 above)
Comment (Jackmin801) [reply src/genesys/utils.py:26]: but the startup will be a few nanoseconds slower and ppl will die!

File: src/genesys/data.py (hunk 1 above)
Comment (Jackmin801) [reply src/genesys/data.py:76]: changed to `do_tokenization`

File: src/genesys/generate.py (hunk 2 above)
Comment (Jackmin801) [reply src/genesys/generate.py:108]: eh should it be global then?

File: src/genesys/data.py [hunk 6]
```diff
@@ -43,7 +43,9 @@ class DataLoaderGenesys:
     Each dataset that is pass must have a "train" split and the content must be a list of dict with at least a "problem" and a "ground_truth" key.
     """
 
-    def __init__(self, config: DataConfig, tokenizer: AutoTokenizer):
+    def __init__(
+        self, config: DataConfig, tokenizer: AutoTokenizer, prime_metric: PrimeMetric, do_tokenization: bool = False
```
Comment (samsja) [inline src/genesys/data.py:47]: should we just do Tokenization = false by default ?

Comment (Jackmin801) [reply src/genesys/data.py:47]: eh that is the default?

Comment (samsja) [reply src/genesys/data.py:47]: sorry I mean, should we just remove this param and just always do =false

Comment (Jackmin801) [reply src/genesys/data.py:47]: Oh but we need the result to be tokenized for toploc. Because encode and decode are not inverses. If we do toploc using the text and we tokenize them, we might get a different token sequence and thus get a false negative.

Comment (Jackmin801) [reply src/genesys/data.py:47]: So now the code returns the detokenized result (bypass detokenizer in sglang). We save the `output_ids` and the `llm_response` in the json. The `llm_response` being the `tokenizer.decode(output_ids)`

Comment (samsja) [reply src/genesys/data.py:47]: My point is that since we pass `do_tokenization=True` I don't see when we use the code to do `do_tokenization=False`

Comment (Jackmin801) [reply src/genesys/data.py:47]: Well if we are pretty sure we also want to tokenized output then sure

**Comment by samsja** [comment]
as discuss tets put toploc in its own repo :) 

PR look good otherwise

File: src/genesys/utils.py [hunk 7]
```diff
@@ -75,7 +76,7 @@ def __del__(self):
             self.worker_thread.join()
 
 
-def save_batch_results(batch_results, results_file, gcp_bucket: GcpBucket | None = None):
+def save_batch_results(batch_results, results_file: "str | Path", gcp_bucket: GcpBucket | None = None):
```
Comment (samsja) [inline src/genesys/utils.py:79]:
```suggestion
def save_batch_results(batch_results, results_file: str | Path, gcp_bucket: GcpBucket | None = None):
```

**Review by samsja — APPROVED** [review]
pre-approving, I added a suggestion tho.

lfgtm
